### PR TITLE
feat(rust,python): Add useful cumulative operations for list series

### PR DESCRIPTION
**Pull request by andysham**
See https://github.com/pola-rs/polars/issues/11504

Our production pipelines process a lot of time series data - one useful pattern here is having a column representing "observed events" at a particular timepoint, which we currently capture using the following `cumulative_eval` call;

```python
lf.with_columns(
    pl.col("events_at_t").cumulative_eval(
        pl.element().drop_nulls().explode().unique().implode()
    ).alias("events_up_to_t")
)
```

This `.drop_nulls().explode().unique().implode()` pattern behaves much like a set union over all lists in the Series. Unfortunately, as said in the documentation for `cumulative_eval` - this is a quadratic operation, which scales quite badly for our needs. This PR provides a linear time implementation of the above expression as `list.cum_set_union`, and similarly provides an expression `list.cum_concat`.

Example usage from the unit tests;
```python
def test_list_cum_concat() -> None:
    a = pl.DataFrame({
        "a": [None, [0], [], None, [1, 5, 9], [0], [], None]
    })
    assert a.select(pl.col("a").list.cum_concat())["a"].to_list() == [
        None,
        [0],
        [0],
        None,
        [0, 1, 5, 9],
        [0, 1, 5, 9, 0],
        [0, 1, 5, 9, 0],
        None,
    ]


def test_list_cum_set_union() -> None:
    a = pl.DataFrame({
        "a": [None, [0], [1, 5], None, [1, 5, 9], [0], [2, 0, 5], None]
    })
    assert a.select(pl.col("a").list.cum_set_union())["a"].to_list() == [
        None,
        [0],
        [0, 1, 5],
        None,
        [0, 1, 5, 9],
        [0, 1, 5, 9],
        [0, 1, 2, 5, 9],
        None,
    ]
```

### CODE REVIEW

File: crates/polars-ops/src/chunked_array/list/cum.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+pub fn list_cum_set_union(a: &ListChunked) -> PolarsResult<ListChunked> {
+    let mut values: Vec<Option<Series>> = Vec::new();
+    let mut acc: Option<Series> = None;
+    for value in a {
+        match (&mut acc, &value) {
+            (Some(acc_inner), Some(v)) => {
+                let mut new_acc = acc_inner.clone();
+                new_acc.append(v)?;
+                new_acc = new_acc.0.unique()?;
```
Comment (ritchie46) [inline crates/polars-ops/src/chunked_array/list/cum.rs:38]: I really don't think this is the way to go. This should be typed and keep a hashtable in `acc`. As it states now, it rebuilds a hash table on every iteration. This is not linear time, but quadratic.

**Review by ritchie46 — CHANGES_REQUESTED** [review]

File: crates/polars-ops/src/chunked_array/list/cum.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-ops/src/chunked_array/list/cum.rs:38]: @andysham maybe polars plugins are a good fit for this?

https://github.com/pola-rs/pyo3-polars

Comment (andysham) [reply crates/polars-ops/src/chunked_array/list/cum.rs:38]: Was not aware of polars plugins - will definitely look into this, thanks!

**Comment by ritchie46** [comment]
I will close this, as I think this one is good fit for polars-plugins.

**Pull request closed without merging**